Repository URL limit increased to 1024 chars

rochacbruno opened the pull request:
* ~~repository auth does not allow utf8 credentials~~ (Allow but needs to be quoted)
* ~~url cannot be longer than 255~~  (Now it is allowing 1024)
* positive and negative auth tests passed locally

rochacbruno commented:
@oshtaier this PR is to fix that issue we talked about yesterday on IRC

oshtaier commented:
ACK, sorry for late response

omaciel commented:
@rochacbruno your comment mentions `repository auth` and `url` but your PR does not touch these 2 fields. Should that have been `login` and `password` instead?

elyezer commented:
Would be great to paste the test results. I'm can't figure out what are the tests related to this fix.

rplevka commented:
NACK


>    repository auth does not allow utf8 credentials
    url cannot be longer than 255

False - such credentials comply with requirements for HTTP authentication thus we want to keep them in our data set.
The trick to use them inside URL is to urlencode them first (see our CLI repository tests)
- the right thing to do is open up a bug for this, not modifying our dataset.

- after you open a bz, you can use `if_bz_open` construct for automation to keep skipping tests with such data until BZ is fixed.

rochacbruno commented:
@elyezer @omaciel in #3359 I added **utf8** data to **valid_http_credentials** then the following test were broken.

```python
py.test -v tests/foreman/api/test_repository.py -k test_negative_create_with_auth
2016-03-31 08:33:40 - nailgun.client - DEBUG - Making HTTP POST request to https://SAT/katello/api/v2/repositories with options {'verify': False, 'auth': ('admin', 'changeme'), 'headers': {'content-type': 'application/json'}} and data {"url": "http://%F0%A9%A8%95%F0%A7%A3%8C%E5%A2%90%EA%A1%9C%E6%AC%8B%F0%92%85%A8%F0%A6%B9%88%F0%A4%A3%85%F0%AA%B4%8C%F0%92%88%A0%E0%AC%9B%F0%A0%A4%AA%EB%B3%84%F0%A2%97%9E%F0%A8%93%AF%F0%A9%82%8C%EB%96%AD%C8%B0%F0%A4%81%97%F0%A3%98%B4%E7%B1%97%E9%A5%B2%F0%A8%A5%99%F0%A0%83%8B%F0%A2%B9%BF%E5%8D%BA%E9%B3%8B%F0%A7%8A%B5%F0%A8%9B%B3%EF%B3%8A%F0%AB%8A%BB%F0%A5%A7%B0%EF%A7%9A%F0%A5%82%BC%F0%9D%99%AB%E9%98%88%F0%A5%A7%8F%F0%92%8D%87%E4%BC%87%EB%92%A9%E2%B1%8E%E9%A0%8A%F0%A8%90%9F%ED%96%96%F0%A6%A0%84%F0%A2%B5%99%E9%BD%94%E7%B0%98%F0%A3%82%8F%E5%B8%A7%F0%AA%94%90%F0%A7%B6%AB%E4%B1%B6%EB%99%86%F0%A1%8F%94%EB%AF%88%F0%A1%BD%91%EC%89%A2%F0%A2%B1%A2%F0%A0%9C%94%E3%BC%83%EB%AA%BD%F0%A8%90%B8%E9%A4%9A%F0%A7%A6%BE%E6%B6%B1%F0%A0%98%82%E6%8B%9F%F0%A4%8F%9E:%F0%A3%A6%B7%F0%A2%89%B3%E7%9A%82%E4%BC%88%E4%A2%95%E7%90%86%F0%AB%9E%B2%F0%AA%B5%86%F0%A1%AB%8D%F0%A7%98%80@rplevka.fedorapeople.org/fakepuppet01/", "product_id": 2, "content_type": "puppet", "name": "\ud856\udfe1\ud842\udf9f\ud835\udf8d\ud85c\udee1"}.
2016-03-31 08:33:41 - nailgun.client - WARNING - Received HTTP 422 response: {"displayMessage":"Validation failed: Url is too long (maximum is 255 characters)","errors":{"url":["is too long (maximum is 255 characters)"]}}
```

By the error **Validation failed: Url is too long (maximum is 255 characters)** I assumed that we can have a user with **utf8*/0-100 chars* but we can't use this user to authenticate against **fedora** repository because leads to a long url.

( @rplevka maybe the solution is passing the auth data in HTTP headers instead url? ) 

So I reverted my changes keeping it as before my previour PR and then tests worked again

```python
py.test -v tests/foreman/api/test_repository.py -k test_negative_create_with_auth
============================================================================ test session starts ============================================================================
platform linux -- Python 3.4.3, pytest-2.9.0, py-1.4.31, pluggy-0.3.1 -- /home/brocha/.virtualenvs/robottelo3/bin/python
cachedir: .cache
rootdir: /home/brocha/Projects/robottelo, inifile: 
plugins: cov-2.2.1, xdist-1.14
collected 36 items 

tests/foreman/api/test_repository.py::RepositoryTestCase::test_negative_create_with_auth_url_too_long PASSED
tests/foreman/api/test_repository.py::RepositoryTestCase::test_negative_create_with_auth_url_with_special_characters PASSED

========================================================= 34 tests deselected by '-ktest_negative_create_with_auth' =========================================================
================================================================= 2 passed, 34 deselected in 64.89 seconds ==================================================================

```

rochacbruno commented:
@rplevka 

> the right thing to do is open up a bug for this (url length), not modifying our dataset.

The length problem happens to authenticate to fedorapeople.org no? so the Bz should be opened for nailgun?

rplevka commented:
@rochacbruno 
let me try to explain my chain of thoughts :)
- normally user is able to use long credentials. Just try to access the fake fedorapeople.org repo and use some of the values from valid credentials list. It works. That means, satellite/katello *should* support them.
As you correctly noted above, Authentication HTTP header is a standard way of passing the credentials allowing to use even long credentials. Unfortunately Katello/Satellite does not support a way of passing the credentials other then including them in the repo URL - and this is should be reported as a bug (as it cannot successfully pass the auth credentials for certain cases even though they're perfectly valid).
Nailgun is innocent in this as it can be only as perfect as Satellite.

rochacbruno commented:
@rplevka did you opened a Bz for that or I should do it?

rplevka commented:
@rochacbruno I haven't please do it. Thanks!

rochacbruno commented:
@rplevka I rerun the tests and suddenly the boundaries has changed.

The error was:

```python
"displayMessage":"Validation failed: Url is too long (maximum is 255 characters)"
```
and now it is 

```python
2016-04-07 10:56:41 - nailgun.client - 
WARNING - Received HTTP 422 response: 
{"displayMessage":"Validation failed: Url is too long (maximum is 1024 characters)"}
```

Now the limit is 1024 chars, so I am going to make the needed changes for the repo tests.

1. valid credentials should have at least one case of 1024 chars url
2. invalid credentials should have at least one case of 1024+ chars url

rochacbruno commented:
@rplevka opened https://bugzilla.redhat.com/show_bug.cgi?id=1325001

Also made some changes in PR:

Changed title to: **Repository URL limit increased to 1024 chars** makes more sense with current status

As at current snap urls can be 1024 chars long I included the data on positive and negative tests adjusting the boudaries to the current limit.

elyezer commented on robottelo/datafactory.py:
```diff
@@ -379,21 +379,21 @@ def valid_http_credentials(url_encoded=False):
         },
         {
-            u'login': gen_string('utf8', gen_integer(1, 100)),
+            u'login': gen_string('utf8', gen_integer(1, 50)),
```
I wasn't able to get why this needs to be 50.

rochacbruno replied:
@elyezer I tested with a varying number to reach an acceptable boundary, assuming that url_enconding will increase the length.

Look this example:

```python
In [23]: len("䑧㭿䑧㭿䑧㭿䑧㭿䑧㭿䑧㭿䑧㭿䑧㭿䑧㭿")
Out[23]: 54

In [24]: len(quote_plus("䑧㭿䑧㭿䑧㭿䑧㭿䑧㭿䑧㭿䑧㭿䑧㭿䑧㭿"))
Out[24]: 162

In [25]: print quote_plus("䑧㭿䑧㭿䑧㭿䑧㭿䑧㭿䑧㭿䑧㭿䑧㭿䑧㭿")
%E4%91%A7%E3%AD%BF%E4%91%A7%E3%AD%BF%E4%91%A7%E3%AD%BF%E4%91%A7%E3%AD%BF%E4%91%A7%E3%AD%BF%E4%91%A7%E3%AD%BF%E4%91%A7%E3%AD%BF%E4%91%A7%E3%AD%BF%E4%91%A7%E3%AD%BF 
```

Having 
150 for user + 150 for password + owner + url + reponame
50 was the safest size I found on tests.

BTW: there is this [RFE] opened https://bugzilla.redhat.com/show_bug.cgi?id=1325001 if able to pass credentials as headers we should pass any value up to 8k (web server HTTP header limit)

elyezer commented:
Can you please update the commit message to reflect the new context?

ACK, pending comments.

oshtaier commented:
Don't like overall idea to play games with numbers and prefer to have solution that is more isolated from app changes, but let's proceed step by step and fix the problem first.

ACK pending comments

oshtaier commented:
I assume that recent changes didn't impact test results
